File: sdks/fastapi/optictrace_fastapi/trace.py
"""W3C trace context for the OpticTrace ASGI middleware.

The point of this module is that correlation must be a fact, not a guess. A
span id generated here is written onto the record AND made available to the
application, so a log line or an outbound call can name the exact request it
belongs to instead of being matched by timestamp — which, under concurrent
traffic, files one tenant's data inside another tenant's request.
"""

from __future__ import annotations

import contextvars
import os
import re
from dataclasses import dataclass
from typing import Optional

HEADER = "traceparent"

# 00-<32 hex trace>-<16 hex span>-<2 hex flags>. Version ff is forbidden by the
# spec; anything malformed starts a fresh trace rather than failing a request —
# losing correlation is a nuisance, failing a request over a bad header is a
# fault.
_TRACEPARENT = re.compile(r"^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$")


@dataclass(frozen=True)
class TraceContext:
    trace_id: str
    span_id: str
    parent_span_id: str = ""
    sampled: bool = True

    def header(self) -> str:
        return f"00-{self.trace_id}-{self.span_id}-{'01' if self.sampled else '00'}"


# The span currently being served. Set by the middleware, read by the log
# handler and by outbound_headers(). A ContextVar is what makes this correct
# under asyncio: every concurrent request gets its own value rather than
# racing over one global.
current: contextvars.ContextVar[Optional[TraceContext]] = contextvars.ContextVar(
    "optictrace_span", default=None
)


def _hex(n: int) -> str:
    return os.urandom(n).hex()


def from_header(raw: str) -> TraceContext:
    """Adopt an inbound traceparent, or start a new trace if there isn't a
    usable one. The caller's span becomes this span's parent, which is what
    makes the result a tree rather than a flat list."""
    m = _TRACEPARENT.match((raw or "").strip())
    if not m:
        return TraceContext(trace_id=_hex(16), span_id=_hex(8))
    version, trace_id, parent, flags = m.groups()
    # All-zero ids are invalid per the spec and mean the caller is broken.
    if version == "ff" or trace_id == "0" * 32 or parent == "0" * 16:
        return TraceContext(trace_id=_hex(16), span_id=_hex(8))
    return TraceContext(
        trace_id=trace_id,
        span_id=_hex(8),
        parent_span_id=parent,
        sampled=bool(int(flags, 16) & 1),
    )


def outbound_headers(extra: Optional[dict] = None) -> dict:
    """Headers to attach to a call this service makes downstream, so the next
    hop nests under this one.

    Carries THIS hop's span, not the caller's. Forwarding the inbound header
    unchanged would make every downstream call a sibling of this request rather
    than a child, and the tree flattens into a list.
    """
    headers = dict(extra or {})
    ctx = current.get()
    if ctx is not None:
        headers[HEADER] = ctx.header()
    return headers

File: sdks/express/trace.js
'use strict';

/**
 * W3C trace context for the Express middleware.
 *
 * Correlation must be a fact, not a guess. The span generated here goes onto
 * the record AND into an AsyncLocalStorage, so a log line or an outbound call
 * can name the exact request it belongs to. Matching on timestamps would,
 * under concurrent traffic, file one tenant's data inside another's request.
 */

const { AsyncLocalStorage } = require('node:async_hooks');
const crypto = require('node:crypto');

const HEADER = 'traceparent';
const TRACEPARENT = /^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/;

// AsyncLocalStorage, not a module-level variable: Node serves requests
// concurrently, and a global would hand one request's span to another.
const storage = new AsyncLocalStorage();

const hex = (bytes) => crypto.randomBytes(bytes).toString('hex');

/**
 * Adopt an inbound traceparent, or start a fresh trace. A malformed header
 * starts a new trace rather than failing anything: losing correlation is a
 * nuisance, failing a request over a bad header would be a fault.
 */
function fromHeader(raw) {
  const m = TRACEPARENT.exec(String(raw ?? '').trim());
  if (m) {
    const [, version, traceId, parent, flags] = m;
    const usable =
      version !== 'ff' && traceId !== '0'.repeat(32) && parent !== '0'.repeat(16);
    if (usable) {
      return {
        traceId,
        spanId: hex(8),
        parentSpanId: parent,
        sampled: (parseInt(flags, 16) & 1) === 1,
      };
    }
  }
  return { traceId: hex(16), spanId: hex(8), parentSpanId: '', sampled: true };
}

const header = (ctx) =>
  `00-${ctx.traceId}-${ctx.spanId}-${ctx.sampled ? '01' : '00'}`;

/** Run `fn` with `ctx` as the current span. */
const run = (ctx, fn) => storage.run(ctx, fn);

/** The span currently being served, or undefined outside a request. */
const current = () => storage.getStore();

/**
 * Headers for a call this service makes downstream, so the next hop nests
 * under this one. Carries THIS hop's span — forwarding the inbound header
 * unchanged would make every downstream call a sibling and flatten the tree.
 */
function outboundHeaders(extra = {}) {
  const ctx = current();
  return ctx ? { ...extra, [HEADER]: header(ctx) } : { ...extra };
}

/** A random id of the given HEX LENGTH — 16 for a span, 32 for a trace.
 *  Exposed because inner spans need ids of the same shape, and a second
 *  generator would be a second place for the convention to drift. */
const randomHex = (hexChars) => hex(hexChars / 2);

module.exports = { HEADER, fromHeader, header, run, current, outboundHeaders, randomHex };
